File: 0714-best-time-to-buy-and-sell-stock-with-transaction-fee/0714-best-time-to-buy-and-sell-stock-with-transaction-fee.cpp
class Solution {
public:
    int maxProfit(vector<int>& p, int fee) {
        // Base case: 
        int profit = 0; 
        int stocks = -p[0]; 
        for (int i = 1; i < p.size(); i++){

            profit = max(profit, stocks + p[i] - fee);
            stocks = max(stocks, profit - p[i]);
        }
        return profit;
    }
};

File: 0714-best-time-to-buy-and-sell-stock-with-transaction-fee/0714-best-time-to-buy-and-sell-stock-with-transaction-fee.java
class Solution:
    def maxProfit(self, prices: List[int], fees: int) -> int:
        dp = {}


        def solve(i, buying):
            if i >= len(prices):
                return 0

            if (i, buying) in dp:
                return dp[(i, buying)]


            
            #means waiting it out and not buying or selling on that day
            cooldown = solve(i + 1, buying)
            if buying:
                buy = solve(i + 1, not buying) - prices[i]
                dp[(i, buying)] = max(buy, cooldown)
            else:
                sell = solve(i + 1, not buying) + prices[i] - fees
                dp[(i, buying)] = max(sell, cooldown)

            return dp[(i, buying)]

        return solve(0, True)